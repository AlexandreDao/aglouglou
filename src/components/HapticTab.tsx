import { BottomTabBarButtonProps } from '@react-navigation/bottom-tabs'
import { PlatformPressable } from '@react-navigation/elements'
import * as Haptics from 'expo-haptics'
import { Platform } from 'react-native'

export const HapticTab = (props: BottomTabBarButtonProps) => {
  return (
    <PlatformPressable
      {...props}
      testID="haptic-tab"
      onPressIn={(ev) => {
        if (Platform.OS === 'ios') {
          // Add a soft haptic feedback when pressing down on the tabs.
          Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light)
        }
        props.onPressIn?.(ev)
      }}
    />
  )
}
