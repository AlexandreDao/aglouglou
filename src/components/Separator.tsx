import { View, StyleSheet } from 'react-native'
import React from 'react'

const styles = StyleSheet.create({
  separator: {
    height: 10,
  },
})

const Separator = () => {
  return <View style={styles.separator} />
}

export default Separator
