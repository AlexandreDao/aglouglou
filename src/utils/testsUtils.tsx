import { render } from '@testing-library/react-native'
import { ReactNode } from 'react'
import { NavigationContainer } from '@react-navigation/native'

export const renderWithNavigation = (children: ReactNode) =>
  render(<NavigationContainer>{children}</NavigationContainer>)
