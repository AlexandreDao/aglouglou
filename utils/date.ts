import { DateFormat } from '@/constants/dateFormat'
import { parse } from 'date-fns'

export const parseDate = (date: string, format = DateFormat.DATE_TIME): Date => {
  return parse(date, format, new Date())
}