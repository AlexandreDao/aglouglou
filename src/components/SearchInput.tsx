import { View, TextInput, StyleSheet } from 'react-native'
import React, { forwardRef, useImperativeHandle, useRef, useState } from 'react'
import { ACTIVE_COLOR, TAB_BAR_BACKGROUND_COLOR, TEXT_COLOR } from '@/constants/colors'
import { IconSymbol } from '@/components/IconSymbol'
import SpeechRecognitionModal from '@/components/SpeechRecognitionModal'
import { Pressable } from 'react-native-gesture-handler'

export interface SearchInputRef {
  focus: () => void
}

interface SearchInputProps {
  onSubmitEditing?: (value: string) => void
  onFocus?: () => void
  onBlur?: () => void
  searchQuery?: string
  setSearchQuery?: (value: string) => void
}

const styles = StyleSheet.create({
  clearBtnContainer: {
    alignItems: 'center',
    borderRadius: 14,
    height: 48,
    justifyContent: 'center',
    width: 48,
  },
  container: {
    alignItems: 'center',
    backgroundColor: TAB_BAR_BACKGROUND_COLOR,
    borderColor: ACTIVE_COLOR,
    borderRadius: 14,
    borderWidth: 2,
    flex: 1,
    flexDirection: 'row',
    marginVertical: 8,
  },
  iconContainer: {
    alignItems: 'center',
    borderColor: ACTIVE_COLOR,
    borderRadius: 24,
    borderWidth: 2,
    height: 48,
    justifyContent: 'center',
    width: 48,
  },
  input: {
    color: TEXT_COLOR,
    flex: 1,
    height: 48,
  },
  inputCleared: {
    paddingHorizontal: 8,
  },
  inputDirty: {
    paddingLeft: 8,
  },
  row: {
    alignItems: 'center',
    flexDirection: 'row',
    gap: 8,
    justifyContent: 'center',
  },
})

const SearchInput = forwardRef<SearchInputRef, SearchInputProps>(
  ({ onSubmitEditing, onFocus, onBlur, searchQuery, setSearchQuery }, ref) => {
    const inputRef = useRef<TextInput>(null)
    const isSearchQueryCleared = !searchQuery?.trim()
    const [isOpen, setIsOpen] = useState(false)

    useImperativeHandle(ref, () => ({
      focus: () => inputRef.current?.focus(),
    }))

    return (
      <View style={styles.row}>
        <View style={styles.container}>
          <TextInput
            testID="search-input"
            ref={inputRef}
            value={searchQuery}
            onFocus={onFocus}
            onBlur={onBlur}
            onSubmitEditing={(e) => {
              onSubmitEditing?.(e?.nativeEvent?.text)
            }}
            submitBehavior="blurAndSubmit"
            onChangeText={setSearchQuery}
            placeholder={'Try typing "Irish"'}
            placeholderTextColor={ACTIVE_COLOR}
            cursorColor={TEXT_COLOR}
            style={[styles.input, isSearchQueryCleared ? styles.inputCleared : styles.inputDirty]}
            enterKeyHint="search"
            inputMode="search"
            returnKeyType="search"
            maxLength={256}
            autoCapitalize="none"
            autoFocus
            enablesReturnKeyAutomatically
          />
          {!isSearchQueryCleared && (
            <Pressable
              style={styles.clearBtnContainer}
              onPress={() => {
                if (inputRef.current) {
                  setSearchQuery?.('')
                  inputRef.current.clear()
                }
              }}
            >
              <IconSymbol name="xmark.circle.fill" color={TEXT_COLOR} />
            </Pressable>
          )}
        </View>
        {isSearchQueryCleared && (
          <Pressable style={styles.iconContainer} onPress={() => setIsOpen(true)}>
            <IconSymbol name="mic" color={TEXT_COLOR} />
          </Pressable>
        )}
        {isOpen && (
          <SpeechRecognitionModal
            isOpen={isOpen}
            setIsOpen={setIsOpen}
            onResult={(value) => {
              setSearchQuery?.(value)
              onSubmitEditing?.(value)
            }}
          />
        )}
      </View>
    )
  }
)

SearchInput.displayName = 'SearchInput'

export default SearchInput
