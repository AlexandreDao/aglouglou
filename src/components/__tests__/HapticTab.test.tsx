import React from 'react'
import { fireEvent } from '@testing-library/react-native'
import { HapticTab } from '@/components/HapticTab'
import * as Haptics from 'expo-haptics'
import { Platform, View } from 'react-native'
import { renderWithNavigation } from '@/utils/testsUtils'

describe('HapticTab', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  test('HapticTab snapshot', () => {
    const mockChildren = <View />
    const tree = renderWithNavigation(<HapticTab>{mockChildren}</HapticTab>).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('triggers haptic feedback on iOS', () => {
    Platform.OS = 'ios'
    const mockChildren = <View />
    const { getByTestId } = renderWithNavigation(<HapticTab>{mockChildren}</HapticTab>)

    fireEvent(getByTestId('haptic-tab'), 'pressIn')
    expect(Haptics.impactAsync).toHaveBeenCalledWith(Haptics.ImpactFeedbackStyle.Light)
  })

  test('does not trigger haptic feedback on Android', () => {
    Platform.OS = 'android'
    const mockChildren = <View />
    const { getByTestId } = renderWithNavigation(<HapticTab>{mockChildren}</HapticTab>)

    fireEvent(getByTestId('haptic-tab'), 'pressIn')
    expect(Haptics.impactAsync).not.toHaveBeenCalled()
  })
})
