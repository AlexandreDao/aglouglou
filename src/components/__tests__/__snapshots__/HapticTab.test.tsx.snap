// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HapticTab HapticTab snapshot 1`] = `
<View
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "cursor": "pointer",
      "opacity": 1,
    }
  }
  testID="haptic-tab"
>
  <View />
</View>
`;
